### PR TITLE
Complete refactoring

### PR DESCRIPTION
**Pull request by PatrykPaziewski**

### CODE REVIEW

File: Dialogs/Second_window_dialog.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import os
+
+from PyQt5 import uic
+from PyQt5 import QtWidgets
+
+
+class Second_window(QtWidgets.QDialog):
```
Comment (bialasjaroslaw) [inline Dialogs/Second_window_dialog.py:7]: Name of that window told me a lot about this class.

File: Dialogs/VX_integration_dialog.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import os
+
+from PyQt5 import uic
+from PyQt5 import QtWidgets
+
+
+class VXDialog(QtWidgets.QDialog):
+    def __init__(self, parent=None):
+        super(VXDialog, self).__init__(parent)
+        uic.loadUi(os.path.join(os.path.dirname(__file__), 'VX_integration_dialog_base.ui'), self)
```
Comment (bialasjaroslaw) [inline Dialogs/VX_integration_dialog.py:10]: Code duplication IMO.
This is the same as previous one. Just initialize from ui and nothing more. Create base class and those two classes should inherit from base. Only difference is name of ui file.

File: Sources/connection.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        self.connection.EntitySelectedInVx += EventHandler(self.drawing.entity_selected_in_vx)
+        self.connection.VxDataCleared += EventHandler(self.drawing.clear_VX_data)
+        self.connection.ReinitializeRequired += EventHandler(self.restart_connection)
+
+    def is_VX_enabled(self):
+        processes = subprocess.Popen('tasklist', stdout=subprocess.PIPE).communicate()[0]
+        if b'WinCanVX.exe' in processes:
+            return True
+        else:
+            return False
```
Comment (bialasjaroslaw) [inline Sources/connection.py:43]:
```suggestion
        return b'WinCanVX.exe' in processes
```

File: Sources/connection.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    def start_connection(self):
+        self.connection.StartCommunication()
+
+    def restart_connection(self):
+        self.stop_connection()
+        self.start_comunication()
+
+    def connect_with_VX(self):
+        conection_timeout = 0
+        while not self.connection.IsConnected and conection_timeout < 10:
```
Comment (bialasjaroslaw) [inline Sources/connection.py:59]: Timeout is usually a measure of time. This one should be called connection_retry_cnt or similar

File: Sources/connection.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            self.start_connection()
+            time.sleep(1)
+            conection_timeout += 1
+        if conection_timeout == 10 and not self.connection.IsConnected:
+            self.parent.show_error(self.parent.tr("Timeout - Connection error after 10 sec. Please make sure that WinCan VX is started."))
```
Comment (bialasjaroslaw) [inline Sources/connection.py:64]: This is not technically true. It could be longer than 10 seconds ;)

File: Sources/connection.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+    def look_for_project(self):
+        project_receive_timeout = 0
+        while not self.is_project() and project_receive_timeout < 10:
+            time.sleep(1)
```
Comment (bialasjaroslaw) [inline Sources/connection.py:69]: sleep at start? Is it ok?

File: Sources/connection.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    def start_comunication(self):
+        self.connected = False
+        self.connect_with_VX()
+        self.look_for_project()        
+        if self.is_project() and self.connection.IsConnected:
+            self.connected = True
+        return self.connected
+
+    def is_project(self):
+        result = type(self.connection.Project) != type(None)
```
Comment (bialasjaroslaw) [inline Sources/connection.py:84]:
```suggestion
        result = self.connection.Project is None
```

File: Sources/drawing.py [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+        self.qgis.mapCanvas().refresh()
+    
+    @pyqtSlot(Model.DataPackage)
+    def redraw(self, data):
+        if not self.layers_created:
+            self.create_layers()
+            
+        if data.IsEmpty:
+            return
+        else:
```
Comment (bialasjaroslaw) [inline Sources/drawing.py:76]: else is not needed after return

File: Sources/drawing.py [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+        self.delete_features(observations, self.created_layers[5])
+        self.delete_features(nodeInspections, self.created_layers[3])
+        self.delete_features(nodeObservations, self.created_layers[4])
+        self.qgis.mapCanvas().refresh()
+                                       
+    def get_selected(self):
+        count = 0
+        selected_shapes = []
+        layer = self.qgis.activeLayer()
+        if type(layer) is not type(None):
```
Comment (bialasjaroslaw) [inline Sources/drawing.py:137]:
```suggestion
        if layer is not None:
```

File: Sources/drawing.py [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+            for feature in layer.selectedFeatures():
+                count += 1
+                selected_shapes.append(feature)
+        return count, selected_shapes, layer
```
Comment (bialasjaroslaw) [inline Sources/drawing.py:141]: so selected_shapes is the same as layer.selectedFeatures() and count is len(selected_shapes), this is so redundant

File: Sources/drawing.py [hunk 11]
```diff
@@ -0,0 +1,418 @@
...
+        inspections = []
+        InspectionLayer = self.created_layers[0]
+        InspectionLayer.startEditing()
+        prov = InspectionLayer.dataProvider()
+        for ins in Inspections:
+            points = []
+            feature = QgsFeature()
+            feature.setFields(self.fields[0])
+            for point in ins.Vertices:
+                points.append(QgsPointXY(point.X, point.Y))
```
Comment (bialasjaroslaw) [inline Sources/drawing.py:154]:
```suggestion
            points = [QgsPointXY(point.X, point.Y) for point in ins.Vertices]
```

Comment (bialasjaroslaw) [reply Sources/drawing.py:154]: then points = [] can be removed

File: Sources/drawing.py [hunk 12]
```diff
@@ -0,0 +1,418 @@
...
+                    for value in self.VX.LayerFieldsPovider.SectionShapeFields:
+                        attr = SectionLayer.fields().indexFromName(str(value.Key))
+                        SectionLayer.changeAttributeValue(
+                            f.id(), attr, str(section.GetValue(value)))
+            else:
+                points = []
+                feature = QgsFeature()
+                feature.setFields(self.fields[1])
+                for point in section.Vertices:
+                    points.append(QgsPointXY(point.X, point.Y))
```
Comment (bialasjaroslaw) [inline Sources/drawing.py:193]: similar to previous comment

File: Sources/drawing.py [hunk 13]
```diff
@@ -0,0 +1,418 @@
...
+
+            if section.ExternalItemId != 0:
+                request = QgsFeatureRequest().setSubsetOfAttributes(
+                    ["OBJ_PK"], SectionLayer.fields()).setFilterExpression('"OBJ_PK"=\'%s\'' % section.Id)
+                features = SectionLayer.getFeatures(request)
+                for f in features:
+                    for value in self.VX.LayerFieldsPovider.SectionShapeFields:
+                        attr = SectionLayer.fields().indexFromName(str(value.Key))
+                        SectionLayer.changeAttributeValue(
+                            f.id(), attr, str(section.GetValue(value)))
```
Comment (bialasjaroslaw) [inline Sources/drawing.py:187]: for/if/for/for I am already lost. Can we make it a function to give a chance to understand what this is doing more expressively?

File: Sources/drawing.py [hunk 14]
```diff
@@ -0,0 +1,418 @@
...
+            else:
+                points = []
+                feature = QgsFeature()
+                feature.setFields(self.fields[1])
+                for point in section.Vertices:
+                    points.append(QgsPointXY(point.X, point.Y))
+                feature.setGeometry(QgsGeometry.fromPolylineXY(points))
+                self.add_values(
+                    feature, section, self.VX.LayerFieldsPovider.SectionShapeFields)
+                sections.append(feature)
```
Comment (bialasjaroslaw) [inline Sources/drawing.py:197]: self.add_values and then modify ots argument. It is not logically ok IMO

File: Sources/drawing.py [hunk 15]
```diff
@@ -0,0 +1,418 @@
...
+            prov.addFeatures(sections)
+            SectionLayer.loadNamedStyle(
+                self.plugin_dir + '\\Styles\\style_section.qml')
+
+        SectionLayer.updateExtents()
+        SectionLayer.commitChanges()
+        QCoreApplication.processEvents()
+
+    def draw_Nodes(self, Nodes):
+        QCoreApplication.processEvents()
```
Comment (bialasjaroslaw) [inline Sources/drawing.py:211]: This is so similar to previous one except nodes instead of sections

File: Sources/drawing.py [hunk 16]
```diff
@@ -0,0 +1,418 @@
...
+            layer_full_path = layers_path + "\\" + layer + ".shp"
+            self.fields.append(self.get_qgis_fields(ShapeFields[nr]))
+            if (layer == node_layer_name or layer == node_observation_layer_name or layer == observation_layer_name):
+                QgsVectorFileWriter(layer_full_path,
+                                    "UTF-8",
+                                    self.get_qgis_fields(ShapeFields[nr]),
+                                    QgsWkbTypes.Point,
+                                    EPSG,
+                                    "ESRI Shapefile")
+            if (layer == inspection_layer_name or layer == node_nspection_layer_name or layer == section_layer_name):
```
Comment (bialasjaroslaw) [inline Sources/drawing.py:367]: almost the same value and action as above

File: Sources/drawing.py [hunk 17]
```diff
@@ -0,0 +1,418 @@
...
+            nr += 1
+
+        QCoreApplication.processEvents()
+        QgsProject.instance().addMapLayers(self.created_layers)
+        self.layers_created = True
+
+    def delete_features(self, to_delete, feature_class):
+        if feature_class is None:
+            return
+        else:
```
Comment (bialasjaroslaw) [inline Sources/drawing.py:385]: no need for else after return

File: Sources/drawing.py [hunk 18]
```diff
@@ -0,0 +1,418 @@
...
+        names = QgsFields()
+        for field in fields:
+            type = self.to_QGIS_type(field.Type)
+            names.append(QgsField(field.Key, type))
+        return names
+
+    def to_QGIS_type(self, VXType):
+        if VXType == 0:
+            return QVariant.Int
+        elif VXType == 1:
```
Comment (bialasjaroslaw) [inline Sources/drawing.py:407]: could also be just simple if

File: Sources/mapping.py [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+            self.wincan_fields_list.addItem(self.mapping_table.item(self.mapping_table.currentRow(), 0).text())
+            self.layer_fields_list.addItem(self.mapping_table.item(self.mapping_table.currentRow(), 1).text())
+            self.mapping_table.removeRow(self.mapping_table.currentRow())
+            
+    def open(self, fields):
+        self.mapped_fields = dict()
+        self.fill_lists_with_field_names(fields)
+        if self.exec_():
+            self.save_mapping_info()
+            return True
```
Comment (bialasjaroslaw) [inline Sources/mapping.py:46]: I would also return false by default

File: Sources/transfer.py [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+        
+        self.vx_fields = Infrastructure.VxFields
+        self.batch_update = CDLAB.WinCan.SDK.GIS.Model.UpdateBatch()
+    
+    def tr(self, message):
+        return QCoreApplication.translate('TransferToWinCan', message)
+
+    def is_wincan_layer(self):
+        if self.layer.name() in self.drawing.wincan_layers:
+            return True
```
Comment (bialasjaroslaw) [inline Sources/transfer.py:41]:
```suggestion
        return self.layer.name() in self.drawing.wincan_layers
```

File: main.py [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+
+    def unload(self):
+        for action in self.actions:
+            self.qgis.removePluginMenu(
+                self.tr(u'&WinCan VX integration'),
+                action)
+        del self.toolbar
+
+        if self.first_start != True:
+            pass
```
Comment (bialasjaroslaw) [inline main.py:157]: Is this needed?

**Review by bialasjaroslaw — CHANGES_REQUESTED** [review]

File: Dialogs/Second_window_dialog.py (hunk 1 above)
Comment (PatrykPaziewski) [reply Dialogs/Second_window_dialog.py:7]: Changed to Mapping_window

File: Dialogs/VX_integration_dialog.py (hunk 2 above)
Comment (PatrykPaziewski) [reply Dialogs/VX_integration_dialog.py:10]: Removed dublication

File: Sources/connection.py (hunk 3 above)
Comment (PatrykPaziewski) [reply Sources/connection.py:43]: Changed

File: Sources/connection.py (hunk 4 above)
Comment (PatrykPaziewski) [reply Sources/connection.py:59]: Changed

File: Sources/connection.py (hunk 6 above)
Comment (PatrykPaziewski) [reply Sources/connection.py:69]: Changed

File: Sources/connection.py (hunk 5 above)
Comment (PatrykPaziewski) [reply Sources/connection.py:64]: Changed

File: Sources/drawing.py (hunk 8 above)
Comment (PatrykPaziewski) [reply Sources/drawing.py:76]: Changed

File: Sources/drawing.py (hunk 9 above)
Comment (PatrykPaziewski) [reply Sources/drawing.py:137]: Changed

File: Sources/drawing.py (hunk 10 above)
Comment (PatrykPaziewski) [reply Sources/drawing.py:141]: Changed

File: Sources/drawing.py (hunk 12 above)
Comment (PatrykPaziewski) [reply Sources/drawing.py:193]: Changed

File: Sources/drawing.py (hunk 13 above)
Comment (PatrykPaziewski) [reply Sources/drawing.py:187]: Changed into method

File: Sources/drawing.py (hunk 16 above)
Comment (PatrykPaziewski) [reply Sources/drawing.py:367]: Changed

File: Sources/drawing.py (hunk 14 above)
Comment (PatrykPaziewski) [reply Sources/drawing.py:197]: Do not get it

File: Sources/drawing.py (hunk 18 above)
Comment (PatrykPaziewski) [reply Sources/drawing.py:407]: Changed

File: Sources/mapping.py (hunk 19 above)
Comment (PatrykPaziewski) [reply Sources/mapping.py:46]: changed

File: Sources/transfer.py (hunk 20 above)
Comment (PatrykPaziewski) [reply Sources/transfer.py:41]: Changed

File: main.py (hunk 21 above)
Comment (PatrykPaziewski) [reply main.py:157]: No - changed

File: Sources/drawing.py (hunk 14 above)
Comment (PatrykPaziewski) [reply Sources/drawing.py:175]: We have discused it.

**Review by bialasjaroslaw — APPROVED** [review]

**Pull request merged**